qadence/ml_tools/train_grad.py

```diff
@@ -332,12 +346,19 @@ def next_loss_iter(dl_iter: Union[None, DataLoader, DictDataLoader]) -> Any:
                 logger.info("Terminating training gracefully after the current iteration.")
                 break
 
-        # Handling printing the last training loss
+        # Handling printing/writing the last training loss
         # as optimize_step does not give the loss value at the last iteration
         try:
             loss, metrics, *_ = next_loss_iter(dl_iter)
             if iteration % config.print_every == 0 and config.verbose:
                 print_metrics(loss, metrics, iteration)
+            write_tracker(
+                writer,
+                loss,
+                metrics,
+                iteration,
+                tracking_tool=config.tracking_tool,
+            )
```

Comment:
Whenever you add a new entry to metrics, the last value is never logged or printed as callbacks are not called here. I would add an additional step where (at least) the custom callbacks are called before `write_tracker`.

Reply:
Ok, I changed that now to go back to having this as part of the final callbacks by specifying a callback `write_tracker` for after optimization, to handle the last loss and metrics.